Fix writing incomplete block

valentin-pinkau opened the pull request:
This PR implements a fix for the case that an incomplete row-major (c order) block is written.
When writing an incomplete block, the current implementation writes the data correctly but fills the rest of the block with scrambled data. The scrambled data origins from the block which is overwritten if it was empty, no harm is done.
The bug comes due to the fact that the existing data is first read in colum-major order into a buffer and the data to write (c order) is loaded in the very same buffer.

I added two tests to catch this bug and implemented a fix by filling the correct `block_buf`.

valentin-pinkau commented:
@amotta filling the buffers with zeros does not solve the problem, so I don't think it is necessary to zero them.

> The bug comes due to the fact that the existing data is first read in colum-major order into a buffer and the data to write (c order) is loaded in the very same buffer.

amotta commented:
@valentin-pinkau Ahh, now I get :) Thank you.

I will try to review the pull request tomorrow.

philippotto commented on python/tests/test_wkw.py:
```diff
@@ -248,6 +248,32 @@ def test_view_on_np_array():
+    assert np.all(read_data[0, :, :, 35:] == 1)
+
+
+def test_not_too_much_data_is_written_in_row_order():
```
The two methods seem to be identical except for the `order="C"` parameter. How about extracting the common code into a helper method which takes the order as a parameter?

philippotto commented on python/tests/test_wkw.py:
```diff
@@ -248,6 +248,32 @@ def test_view_on_np_array():
+    data = generate_test_data(np.uint8, data_shape, order="C")
+    with wkw.Dataset.create("tests/tmp", wkw.Header(np.uint8)) as dataset:
+        dataset.write((0, 0, 0), np.ones((35, 35, 64), dtype=np.uint8))
+        dataset.write((0, 0, 0), data)
```
I'd write the data at `(1, 2, 3)` instead of `(0, 0, 0)` to also check the partial-write at the "top left" side of the block (right now, only the "bottom-right" side is tested).
Then, another assertion should be added which mirrors the semantics of the `assert np.all(read_data[0, :, :, 35:] == 1)` assertion.

valentin-pinkau commented:
@philippotto @amotta Thank you for your reviews! Please have another look at the code. 
I refactored the `File::write_mat` method and moved the order handling into the `copy_from_order_agnostic` method. I like the code much better now.
I added and refactored the tests based on your feedback.
